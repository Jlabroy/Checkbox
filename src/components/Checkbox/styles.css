.checkbox {
    cursor: pointer;
}

    .checkbox .checkbox__box {
        position: relative;
        display: inline-block;
        vertical-align: top;
        width: 15px;
        height: 15px;
        border: 1px solid #ccc;
        border-radius: 3px;
    }

        .checkbox.checkbox--state-active .checkbox__box {
            background-color: blue;
            border-color: blue;
        }

            .checkbox.checkbox--state-active .checkbox__box::after {
                content: '';
                position: absolute;
                left: 2px;
                top: 7px;
                background: white;
                width: 2px;
                height: 2px;
                box-shadow:
                        2px 0 0 white,
                        4px 0 0 white,
                        4px -2px 0 white,
                        4px -4px 0 white,
                        4px -6px 0 white,
                        4px -8px 0 white;
                transform: rotate(45deg);
            }

        .checkbox.checkbox--type-disabled .checkbox__box {
            background-color: #ccc;
            border-color: #ccc;
        }

    .checkbox .checkbox__label {
        display: inline-block;
        vertical-align: top;
        margin-left: 10px;
    }