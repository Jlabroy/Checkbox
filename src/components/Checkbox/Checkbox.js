import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import "./styles.css";

class Checkbox extends PureComponent {
  static propTypes = {
    /** Whether or not the checkbox is checked. */
    isChecked: PropTypes.bool,

    /** Whether or not the checkbox is disabled. */
    isDisabled: PropTypes.bool,

    /**
     * Optional function called when the user changes the checked state.
     * If this function is not null then the component becomes controlled
     * and will not manage its own state.
     *
     * @param {string|number}  value  The value of the checkbox being changed.
     * @param {boolean}  isChecked  Whether the checkbox is checked.
     */
    onChange: PropTypes.func,

    /** The checkboxs value. */
    value: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number
    ]),

    /** Optional class name passed to the checkbox. */
    className: PropTypes.string,

    /** Text string of label with the checkbox. */
    label: PropTypes.string
  };

  static defaultProps = {
    isDisabled: false,
    onChange: null,
    value: "",
    className: "",
    label: ""
  };

  static getDerivedStateFromProps({ isChecked }) {
    if (!isChecked) {
      return {};
    }

    return {
      isChecked
    };
  }

  state = {
    isChecked: false
  };

  handleClick = () => {
    const { isChecked } = this.state;
    const { onChange, value, isDisabled } = this.props;

    if (isDisabled) {
      return false;
    }

    if (onChange) {
      onChange(value, !isChecked);
      return;
    }

    this.setState(({ isChecked }) => ({
      isChecked: !isChecked
    }));
  };

  render() {
    const { isChecked } = this.state;
    const { label, className, isDisabled } = this.props;

    return (
      <div
        onClick={this.handleClick}
        className={`
          checkbox
          checkbox--state-${isChecked ? "active" : "inactive"}
          checkbox--type-${isDisabled ? "disabled" : "enabled"}
          ${className}
        `}
      >
        <div className="checkbox__box" />
        {label && (
          <div className="checkbox__label">
            {label}
          </div>
        )}
      </div>
    );
  }
}

export default Checkbox;
