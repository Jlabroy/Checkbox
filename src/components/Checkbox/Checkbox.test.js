import React from "react";
import ReactDOM from "react-dom";
import Enzyme, { shallow } from "enzyme";
import EnzymeAdapter from "enzyme-adapter-react-16";
import toJson from "enzyme-to-json";
import Checkbox from "./Checkbox";

Enzyme.configure({ adapter: new EnzymeAdapter() });

describe("Checkbox test suite", () => {
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<Checkbox />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it("matches snapshot", () => {
    const Component = <Checkbox />;
    expect(toJson(shallow(Component))).toMatchSnapshot();
  });
  
  it("matches snapshot - default checked", () => {
    const Component = <Checkbox isChecked />;
    expect(toJson(shallow(Component))).toMatchSnapshot();
  });

  it("matches snapshot - default disabled", () => {
    const Component = <Checkbox isDisabled />;
    expect(toJson(shallow(Component))).toMatchSnapshot();
  });

  it("matches snapshot - additional classes", () => {
    const Component = <Checkbox className="additional" />;
    expect(toJson(shallow(Component))).toMatchSnapshot();
  });

  it("matches snapshot - with label", () => {
    const Component = <Checkbox label="Test label" />;
    expect(toJson(shallow(Component))).toMatchSnapshot();
  });

  it("matches snapshot - before and after click", () => {
    const Component = shallow(<Checkbox label="Test label" />);

    Component.simulate("click");
    expect(toJson(Component)).toMatchSnapshot();
  });

  it("Can handle custom onChange functions", () => {
    const onChange = jest.fn();
    const testValue = "test value";

    const Component = shallow(
      <Checkbox
        onChange={onChange}
        value={testValue}
      />
    );

    Component.simulate("click");
    expect(onChange).toHaveBeenCalledWith(testValue, true);
  });

  it("Doesn't run custom onChange if disabled", () => {
    const onChange = jest.fn();

    const Component = shallow(
      <Checkbox
        isDisabled
        onChange={onChange}
        value="test value"
      />
    );

    Component.simulate("click");
    expect(onChange).not.toHaveBeenCalled();
  });

  it("Doesn't update if disabled when clicked", () => {
    const Component = shallow(<Checkbox isDisabled />);

    Component.simulate("click");
    expect(toJson(Component)).toMatchSnapshot();
  });
});
