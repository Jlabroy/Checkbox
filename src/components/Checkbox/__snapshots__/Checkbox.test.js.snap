// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox test suite Doesn't update if disabled when clicked 1`] = `
<div
  className="checkbox checkbox--state-inactive checkbox--type-disabled"
  onClick={[Function]}
>
  <div
    className="checkbox__box"
  />
</div>
`;

exports[`Checkbox test suite matches snapshot - additional classes 1`] = `
<div
  className="checkbox checkbox--state-inactive checkbox--type-enabled additional"
  onClick={[Function]}
>
  <div
    className="checkbox__box"
  />
</div>
`;

exports[`Checkbox test suite matches snapshot - before and after click 1`] = `
<div
  className="checkbox checkbox--state-active checkbox--type-enabled"
  onClick={[Function]}
>
  <div
    className="checkbox__box"
  />
  <div
    className="checkbox__label"
  >
    Test label
  </div>
</div>
`;

exports[`Checkbox test suite matches snapshot - default checked 1`] = `
<div
  className="checkbox checkbox--state-active checkbox--type-enabled"
  onClick={[Function]}
>
  <div
    className="checkbox__box"
  />
</div>
`;

exports[`Checkbox test suite matches snapshot - default disabled 1`] = `
<div
  className="checkbox checkbox--state-inactive checkbox--type-disabled"
  onClick={[Function]}
>
  <div
    className="checkbox__box"
  />
</div>
`;

exports[`Checkbox test suite matches snapshot - with label 1`] = `
<div
  className="checkbox checkbox--state-inactive checkbox--type-enabled"
  onClick={[Function]}
>
  <div
    className="checkbox__box"
  />
  <div
    className="checkbox__label"
  >
    Test label
  </div>
</div>
`;

exports[`Checkbox test suite matches snapshot 1`] = `
<div
  className="checkbox checkbox--state-inactive checkbox--type-enabled"
  onClick={[Function]}
>
  <div
    className="checkbox__box"
  />
</div>
`;
